packages/plugin-ext/src/main/browser/custom-editors/custom-editors-main.ts

```diff
@@ -0,0 +1,543 @@
+            if (this.ongoingSave === cancelable) { // Make sure we are still doing the same save
+                this.ongoingSave = undefined;
+            }
+        }
+    }
+
+    async saveAs(resource: TheiaURI, targetResource: TheiaURI, options?: CopyFileOptions): Promise<void> {
+        if (this.editable) {
+            const source = new CancellationTokenSource();
+            await this.proxy.$onSaveAs(this.resource, this.viewType, URI.file(targetResource.path.toString()), source.token);
+            this.change(() => {
+                this.savePoint = this.currentEditIndex;
+            });
+        } else {
+            // Since the editor is readonly, just copy the file over
+            await this.fileService.copy(resource, targetResource, { overwrite: false });
+        }
+    }
+}
+
+// copied from https://github.com/microsoft/vscode/blob/53eac52308c4611000a171cc7bf1214293473c78/src/vs/workbench/contrib/customEditor/common/customTextEditorModel.ts
+export class CustomTextEditorModel implements CustomEditorModel {
+
+    private readonly toDispose = new DisposableCollection();
+    readonly autoSave: 'on' | 'off';
+
+    private readonly onDirtyChangedEmitter = new Emitter<void>();
+    readonly onDirtyChanged = this.onDirtyChangedEmitter.event;
+
+    static async create(
+        viewType: string,
+        resource: TheiaURI,
+        editorModelService: EditorModelService,
+        fileService: FileService
+    ): Promise<CustomTextEditorModel> {
+        const model = await editorModelService.createModelReference(resource);
+        return new CustomTextEditorModel(viewType, resource, model, fileService);
+    }
+
+    constructor(
+        readonly viewType: string,
+        readonly editorResource: TheiaURI,
+        private readonly model: Reference<MonacoEditorModel>,
+        private readonly fileService: FileService
+    ) {
+        this.toDispose.push(
+            this.editorTextModel.onDirtyChanged(e => {
+                this.onDirtyChangedEmitter.fire();
+            })
+        );
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+        this.model.dispose();
+    }
+
+    get resource(): URI {
+        return URI.file(this.editorResource.path.toString());
+    }
+
+    get dirty(): boolean {
+        return this.editorTextModel.dirty;
+    };
+
+    get readonly(): boolean {
+        return this.editorTextModel.readOnly;
+    }
+
+    get editorTextModel(): MonacoEditorModel {
+        return this.model.object;
+    }
+
+    revert(options?: Saveable.RevertOptions): Promise<void> {
+        return this.editorTextModel.revert(options);
+    }
+
+    save(options?: SaveOptions): Promise<void> {
+        return this.editorTextModel.save(options);
+    }
+
+    async saveAs(resource: TheiaURI, targetResource: TheiaURI, options?: CopyFileOptions): Promise<void> {
+        await this.save({ formatType: FormatType.ON });
+        await this.fileService.copy(resource, targetResource, options);
+    }
```

Comment:
Do I understand correctly that this will override the original file, as well as create the new one by doing a copy?

This behavior is not what one would expect from a `save as` feature right? I would expect to only create a new file with the new changes, leaving the old one untouched?

Reply:
@marechal-p 
As I didn't start this PR, after checking this, it seems that this logic is same logic implemented for saveAs command = save file and then copy:
![image](https://user-images.githubusercontent.com/71069170/106114699-887f6a00-6158-11eb-982a-a9dcfde5a264.png)